pkg/bff/db/announcements.go

```diff
@@ -0,0 +1,215 @@
+	// and ordering mechanism or implement reverse seek in trtl.
+	results := make([]*api.Announcement, 0)
+	for {
+		var pair *trtl.KVPair
+		if pair, err = cursor.Recv(); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("could not read message from cursor: %s", err)
+		}
+
+		if bytes.Equal(pair.Key, lastUpdatedKey) {
+			continue
+		}
+
+		var post *api.Announcement
+		if post, err = a.Decode(pair.Value); err != nil {
+			// If we can't decode the announcement log the error and continue
+			log.Error().Err(err).Str("key", ParseKSUID(pair.Key)).Msg("could not decode announcement from database")
+			continue
+		}
+
+		var date time.Time
+		if date, err = time.Parse("2006-01-02", post.PostDate); err != nil {
+			// If we can't decode the post date log the error and continue
+			log.Error().Err(err).Str("key", ParseKSUID(pair.Key)).Msg("could not parse the post date")
+		}
+
+		if date.Before(notBefore) {
+			// Filter on the not before date
+			continue
+		}
+
+		results = append(results, post)
+	}
+
+	// Reverse the results and apply the max results limit
+	for i := len(results) - 1; i >= 0; i-- {
+		if len(out.Announcements) >= maxResults {
+			break
+		}
+		out.Announcements = append(out.Announcements, results[i])
+	}
+	return out, nil
+}
+
+// Post an announcement, putting it to the trtl database. This method does no
+// verification of duplicate announcements or any content verification except for a
+// check that an empty announcement is not being put to the database.
+func (a *Announcements) Post(ctx context.Context, in *api.Announcement) (_ string, err error) {
+	// Make sure we don't post empty announcements
+	if in.Title == "" && in.Body == "" && in.PostDate == "" && in.Author == "" {
+		return "", ErrEmptyAnnouncement
+	}
+
+	// Encode the Post to bytes for storage
+	var value []byte
+	if value, err = a.Encode(in); err != nil {
+		return "", fmt.Errorf("could not encode announcement: %s", err)
+	}
+
+	// Create a ksuid for the post
+	key := ksuid.New()
+
+	if err = a.db.Put(ctx, key.Bytes(), value, a.namespace); err != nil {
+		return "", fmt.Errorf("could not put announcement to db: %s", err)
+	}
+
+	// Set the last updated timestamp on the database
+	var updated []byte
+	if updated, err = json.Marshal(time.Now()); err != nil {
+		return "", fmt.Errorf("could not create last updated timestamp value: %s", err)
+	}
+
+	if err = a.db.Put(ctx, lastUpdatedKey, updated, a.namespace); err != nil {
+		return "", fmt.Errorf("could not put last updated index: %s", err)
+	}
+
+	return key.String(), nil
+}
+
+// Namespace implements the collection interface
+func (a *Announcements) Namespace() string {
+	return a.namespace
+}
+
+// Encode an announcement for storage in the database.
+// HACK: the api definition is probably not the right place to store a database model,
+// should we just create protocol buffer model definitions instead of using this?
```

Comment:
Here's a (maybe terrible) idea, if we are hard limiting the number of announcements can we just maintain the entire ordered list in a single object in the database? It opposes the way we are managing the other records but if we keep the size restricted then it simplifies the handling.

Reply:
That's a very similar idea to storying a list of announcements per month - and I think there is definitely something to it!